exercises/robot-simulator/example.rs

```diff
@@ -0,0 +1,102 @@
+#[derive(PartialEq, Debug, Clone)]
+pub enum Direction {
+    North,
+    East,
+    South,
+    West,
+}
+
+impl Direction {
+    pub fn previous_clockwise(&self) -> Self {
+        match self {
+            &Direction::North => Direction::West,
+            &Direction::East => Direction::North,
+            &Direction::South => Direction::East,
+            &Direction::West => Direction::South,
+        }
+    }
+
+    pub fn next_clockwise(&self) -> Self {
+        match self {
+            &Direction::North => Direction::East,
+            &Direction::East => Direction::South,
+            &Direction::South => Direction::West,
+            &Direction::West => Direction::North,
+        }
+    }
+}
+
+struct Position {
+    x: i32,
+    y: i32,
+}
+
+impl Position {
+    fn new(x: i32, y: i32) -> Self {
+        Position { x: x, y: y }
+    }
+
+    fn advance(self, direction: &Direction) -> Self {
+        match direction {
+            &Direction::North => Self::new(self.x, self.y + 1),
+            &Direction::South => Self::new(self.x, self.y - 1),
+            &Direction::East => Self::new(self.x + 1, self.y),
+            &Direction::West => Self::new(self.x - 1, self.y),
+        }
+    }
+}
+
+pub struct Robot {
+    position: Position,
+    direction: Direction,
+}
+
+impl Robot {
+    pub fn new(x: i32, y: i32, d: &Direction) -> Self {
+        Robot::build(Position::new(x, y), d.clone())
+    }
+
+    fn build(position: Position, direction: Direction) -> Self {
+        Robot {
+            position: position,
+            direction: direction,
+        }
+    }
+
+    pub fn turn_right(self) -> Self {
+        Self::build(self.position, self.direction.next_clockwise())
+    }
+
+    pub fn turn_left(self) -> Self {
+        Self::build(self.position, self.direction.previous_clockwise())
+    }
+
+    pub fn advance(self) -> Self {
+        Self::build(self.position.advance(&self.direction), self.direction)
+    }
+
+    pub fn instructions(self, instructions: &str) -> Self {
+        let mut r = Self::build(self.position, self.direction);
+        for c in instructions.chars() {
```

Comment:
well, looks like you did want to use `fold`, and well I'll not complain about you using it! I think I only hesitate because I'm not sure if the typical Rust user is comfortable thinking in `fold`s, so I'm never sure whether it's idiomatic and whether I should just use a (relatively more) straightforward `for` instead. Maybe I should have more faith. It is true that it can result in shorter code.